client/service/ledger_validation.go

```diff
@@ -151,15 +148,21 @@ func (s ClientService) ValidateLedger(args ...interface{}) (result model.LedgerV
 
 			r <- response
 			e <- err
-			close(r)
-			close(e)
 		}(ledgerChan, ledgerErrorChan)
 
-		if responseFromLedger, err = <-ledgerChan, <-ledgerErrorChan; err != nil {
+		var errFromLedger error
+		responseFromLedger, errFromLedger = <-ledgerChan, <-ledgerErrorChan
+
+		var errFromAuditor error
+		responseFromAuditor, err = <-auditorChan, <-auditorErrorChan
+
+		if errFromLedger != nil {
+			err = errFromLedger
 			return
 		}
 
-		if responseFromAuditor, err = <-auditorChan, <-auditorErrorChan; err != nil {
+		if errFromAuditor != nil {
+			err = errFromAuditor
```

Comment:
This part is a bit hard to understand.

Imaging that if there's an error occurs and returns from the ledger's GO routine, the original logic will get an error and just return the whole function which forcibly closes all channels because of `defer` statements, then, the auditor's GO routine might send the value to a nil channel.

So, wait until all GO routines finish then do the return statement.